src/Nethermind/Nethermind.Specs/ChainSpecStyle/ZstdChainSpecLoader.cs

```diff
@@ -0,0 +1,43 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.IO;
+using System.Reflection;
+using Nethermind.Config;
+using ZstdSharp;
+
+namespace Nethermind.Specs.ChainSpecStyle;
+
+public class ZstdChainSpecLoader : IChainSpecLoader
+{
+    private const string EmbeddedDictionaryPath = "Nethermind.Config.chainspec.dictionary";
+
+    private readonly string? _dictionaryPath;
+    private readonly IChainSpecLoader _decompressedLoader;
+
+    public ZstdChainSpecLoader(IChainSpecLoader decompressedLoader, string? dictionaryPath = null)
+    {
+        _decompressedLoader = decompressedLoader;
+        _dictionaryPath = dictionaryPath;
+    }
+
+    public ChainSpec Load(Stream streamData)
+    {
+        using var buffer = new MemoryStream();
+        if (_dictionaryPath is null)
+        {
+            Assembly assembly = typeof(IConfig).Assembly;
+            using Stream stream = assembly.GetManifestResourceStream(EmbeddedDictionaryPath)!;
+            stream.CopyTo(buffer);
+        }
+        else
+        {
+            using Stream stream = File.OpenRead(_dictionaryPath);
+            stream.CopyTo(buffer);
+        }
+
+        using var decompressedStream = new DecompressionStream(streamData);
+        decompressedStream.LoadDictionary(buffer.ToArray());
```

Comment:
Unfortunately `LoadDictionary` does not accept any async/Stream overload.
In practice it should not be an issue since the `dictionary` is ~3.5 Mb and it's only loaded once during initialization.

Reply:
Ok but you first load it to MemoryStream and then ToArray, can you just use one data in the first place?

Reply:
You mean that each `Stream` inside the conditional produce a `byte[]` directly instead of going through `MemoryStream`? If so then I might be missing some API because I'm not sure how to do that.

Reply:
```
        using Stream stream = _dictionaryPath is null
            ? typeof(IConfig).Assembly.GetManifestResourceStream(EmbeddedDictionaryPath)!
            : File.OpenRead(_dictionaryPath);

        byte[] buffer = new byte[stream.Length];
        stream.ReadExactly(buffer);

        using var decompressedStream = new DecompressionStream(streamData);
        decompressedStream.LoadDictionary(buffer);
```

Reply:
Didn't know that you could "just ask" for a Stream length. Changed.